### PR TITLE
 mrc-3028 basic location removal

### PR DESCRIPTION
**Pull request by hillalex**
This does the minimal thing of removing a location from the config and marking packets from this location as now belonging to the "orphan" location instead. 

The ticket mentions an option to prune metadata instead of orphaning it. I guess in this case we'd also remove unpacked packets. Otherwise, leaving unpacked packets as they are seems fine to me?

### CODE REVIEW

File: R/location.R [hunk 1]
```diff
@@ -95,6 +95,55 @@ outpack_location_rename <- function(old, new, root = NULL) {
...
+        config$location,
+        new_location_entry(orphan, -1, "orphan", NULL))
+      config$location <- config$location[
+        order(config$location$priority, decreasing = TRUE), ]
+      rownames(config$location) <- NULL
+    }
+
+    orphan_id <- config$location$id[match("orphan", config$location$name)]
+    mark_packets_orphaned(id, packets, orphan_id, root)
+    root$index(refresh = TRUE)
```
Comment (hillalex) [inline R/location.R]: is this the way to do this?

Comment (richfitz) [reply R/location.R]: yeah, the workflow here is not really obvious yet, I think that most times we load the index there's a refresh anyway, but at some point we should look at this holistically and make sure it's always done when we expect it to be done

File: R/location.R [hunk 2]
```diff
@@ -95,6 +95,55 @@ outpack_location_rename <- function(old, new, root = NULL) {
...
+      config$location <- rbind(
+        config$location,
+        new_location_entry(orphan, -1, "orphan", NULL))
+      config$location <- config$location[
+        order(config$location$priority, decreasing = TRUE), ]
+      rownames(config$location) <- NULL
+    }
+
+    orphan_id <- config$location$id[match("orphan", config$location$name)]
+    mark_packets_orphaned(id, packets, orphan_id, root)
```
Comment (richfitz) [inline R/location.R]: This should only orphan packets that are *only* found at the removed location, but I think this is orphaning any packets that are found there, even if they are also found elsewhere.

So imagine we've pulled packages from production but also some test server. We remove the test server and we should be orphaning (or removing) only the packages found only from the test server

File: R/location.R [hunk 3]
```diff
@@ -95,6 +95,55 @@ outpack_location_rename <- function(old, new, root = NULL) {
...
+    if (!location_exists(root, "orphan")) {
+      config$location <- rbind(
+        config$location,
+        new_location_entry(orphan, -1, "orphan", NULL))
+      config$location <- config$location[
+        order(config$location$priority, decreasing = TRUE), ]
+      rownames(config$location) <- NULL
+    }
+
+    orphan_id <- config$location$id[match("orphan", config$location$name)]
```
Comment (richfitz) [inline R/location.R:132]: lookup_location_id here

File: R/location.R [hunk 4]
```diff
@@ -95,6 +95,55 @@ outpack_location_rename <- function(old, new, root = NULL) {
...
+    mark_packets_orphaned(id, packets, orphan_id, root)
+    root$index(refresh = TRUE)
+  }
+
+  config$location <- config$location[config$location$name != name, ]
+  config_write(config, root$path)
+
+  root$config <- config_read(root$path)
```
Comment (richfitz) [inline R/location.R]: We must do this pattern (write then read) a few places - might be worth pulling it into a small function in config.R (config_update <- function(config, root) {
  config_write(config, root$path)
  root$config <- config_read(root$path)
  })

or a similar method on the root object?

**Review by richfitz — CHANGES_REQUESTED** [review]
I am not sure about the deletion vs orphan difference - something to discuss perhaps? Doesn't need to be done here but I think that deletion does make some sense if we can do it because once we can no longer pull or interact with the packets we might want to be able to remove them.

Another way of doing this would be to have some sort of housekeeping "delete orphans" function that we can call later?

File: R/location.R (hunk 3 above)
Comment (hillalex) [reply R/location.R:132]: at this point `root$config` might not have a newly created orphan location

**Review by richfitz — APPROVED** [review]

**Pull request merged**